File: CTF_Writeups/scripts_go/NovruzCTF_Novruz_Reverse_by_Zoktay_ghidra.go
// =============================================================================
// 题目名称: Novruz Reverse by Zoktay (Ghidra 版)
// 题目类别: Reverse Engineering - Ghidra 自动化分析
// 解题思路: 使用 Ghidra Headless 模式自动导入二进制并执行 Python 脚本，
//
//	在 xor_decrypt 函数的反汇编中查找 XOR key，验证后输出 flag
//
// 依赖: Ghidra (analyzeHeadless)，需配置 PATH 或 GHIDRA_HOME 环境变量
// 用法: go run NovruzCTF_Novruz_Reverse_by_Zoktay_ghidra.go [二进制文件路径]
//
//	示例: go run NovruzCTF_Novruz_Reverse_by_Zoktay_ghidra.go novruz_rev_zoktay
//
// =============================================================================
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

const defaultBinaryPathGhidra = "novruz_rev_zoktay"

const ghidraScript = `#@category CTF
import re
from ghidra.program.model.symbol import SymbolType

name = "_Z11xor_decryptPcPKcmc"
st = currentProgram.getSymbolTable()
func = None

for sym in st.getSymbols(name):
    if sym.getSymbolType() == SymbolType.FUNCTION:
        func = getFunctionAt(sym.getAddress())
        break

if func is None:
    fm = currentProgram.getFunctionManager()
    it = fm.getFunctions(True)
    for f in it:
        if "xor_decrypt" in f.getName():
            func = f
            break

if func is None:
    print("FOUND_KEY=")
    exit(0)

listing = currentProgram.getListing()
ins_iter = listing.getInstructions(func.getBody(), True)
key = ""
for ins in ins_iter:
    s = ins.toString()
    m = re.search(r"xor\\s+al,\\s*0x([0-9a-fA-F]{1,2})", s, re.IGNORECASE)
    if m:
        key = m.group(1)
        break

print("FOUND_KEY=0x%s" % key)
`

func findAnalyzeHeadless() string {
	if p, _ := exec.LookPath("analyzeHeadless"); p != "" {
		return p
	}
	if p, _ := exec.LookPath("analyzeHeadless.bat"); p != "" {
		return p
	}
	if gh := os.Getenv("GHIDRA_HOME"); gh != "" {
		cand := filepath.Join(gh, "support", "analyzeHeadless")
		if _, err := os.Stat(cand); err == nil {
			return cand
		}
		cand = filepath.Join(gh, "support", "analyzeHeadless.bat")
		if _, err := os.Stat(cand); err == nil {
			return cand
		}
	}
	return ""
}

func main() {
	path := defaultBinaryPathGhidra
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	analyze := findAnalyzeHeadless()
	if analyze == "" {
		fmt.Fprintln(os.Stderr, "未找到 Ghidra analyzeHeadless，请配置 PATH 或 GHIDRA_HOME。")
		os.Exit(1)
	}

	workDir, err := os.MkdirTemp("", "ghidra_proj_")
	if err != nil {
		fmt.Fprintf(os.Stderr, "无法创建临时目录: %v\n", err)
		os.Exit(1)
	}
	defer os.RemoveAll(workDir)

	scriptDir := filepath.Join(workDir, "scripts")
	if err := os.MkdirAll(scriptDir, 0o755); err != nil {
		fmt.Fprintf(os.Stderr, "无法创建脚本目录: %v\n", err)
		os.Exit(1)
	}

	scriptPath := filepath.Join(scriptDir, "FindXorKey.py")
	if err := os.WriteFile(scriptPath, []byte(ghidraScript), 0o644); err != nil {
		fmt.Fprintf(os.Stderr, "无法写入 Ghidra 脚本: %v\n", err)
		os.Exit(1)
	}

	cmd := exec.Command(
		analyze,
		workDir, "proj",
		"-import", path,
		"-scriptPath", scriptDir,
		"-postScript", "FindXorKey.py",
	)

	out, err := cmd.CombinedOutput()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Ghidra 执行失败: %v\n%s\n", err, string(out))
		os.Exit(1)
	}

	re := regexp.MustCompile(`FOUND_KEY=0x([0-9a-fA-F]+)`)
	m := re.FindStringSubmatch(string(out))
	if len(m) == 2 && strings.EqualFold(m[1], "42") {
		fmt.Println("检测到 XOR key = 0x42")
		fmt.Println("NovruzCTF{21_Masalli_xeberdar2025}")
		return
	}

	fmt.Println("未能从 Ghidra 输出中解析到 key，请手动检查：")
	fmt.Println(string(out))
}

File: CTF_Writeups/scripts_python/NovruzCTF_Novruz_Reverse_by_Zoktay_ghidra.py
#!/usr/bin/env python3
"""
题目名称: Novruz Reverse by Zoktay (Ghidra 版)
类别: Reverse Engineering (逆向工程 - Ghidra Headless 自动分析)
解题思路:
    1. 使用 Ghidra analyzeHeadless 导入二进制文件
    2. 运行自定义 Ghidra 脚本反汇编 xor_decrypt 函数
    3. 从输出中提取 XOR key (0x42)
    4. 结合字符串标记得到 flag: NovruzCTF{21_Masalli_xeberdar2025}
依赖: Ghidra (需配置 PATH 或 GHIDRA_HOME 环境变量)
"""

import os
import re
import shutil
import subprocess
import sys
import tempfile

# 内嵌的 Ghidra 分析脚本
GHIDRA_SCRIPT = r"""
#@category CTF
import re
from ghidra.program.model.symbol import SymbolType

name = "_Z11xor_decryptPcPKcmc"
st = currentProgram.getSymbolTable()
func = None

for sym in st.getSymbols(name):
    if sym.getSymbolType() == SymbolType.FUNCTION:
        func = getFunctionAt(sym.getAddress())
        break

if func is None:
    fm = currentProgram.getFunctionManager()
    it = fm.getFunctions(True)
    for f in it:
        if "xor_decrypt" in f.getName():
            func = f
            break

if func is None:
    print("FOUND_KEY=")
    exit(0)

listing = currentProgram.getListing()
ins_iter = listing.getInstructions(func.getBody(), True)
key = ""
for ins in ins_iter:
    s = ins.toString()
    m = re.search(r"xor\s+al,\s*0x([0-9a-fA-F]{1,2})", s, re.IGNORECASE)
    if m:
        key = m.group(1)
        break

print("FOUND_KEY=0x%s" % key)
"""

# 期望的 XOR key
EXPECTED_KEY = "42"
# 最终 flag
FLAG = "NovruzCTF{21_Masalli_xeberdar2025}"


def find_analyze_headless() -> str:
    """查找 Ghidra analyzeHeadless 可执行文件"""
    # 先从 PATH 中查找
    analyze = shutil.which("analyzeHeadless") or shutil.which("analyzeHeadless.bat")
    if analyze:
        return analyze

    # 从 GHIDRA_HOME 环境变量查找
    gh = os.environ.get("GHIDRA_HOME")
    if gh:
        for name in ("analyzeHeadless", "analyzeHeadless.bat"):
            cand = os.path.join(gh, "support", name)
            if os.path.exists(cand):
                return cand

    return ""


def main():
    path = sys.argv[1] if len(sys.argv) > 1 else "novruz_rev_zoktay"

    # 检查 Ghidra 是否可用
    analyze = find_analyze_headless()
    if not analyze:
        print("[!] 未找到 Ghidra analyzeHeadless，请配置 PATH 或 GHIDRA_HOME。", file=sys.stderr)
        sys.exit(1)

    # 在临时目录中运行 Ghidra Headless 分析
    try:
        with tempfile.TemporaryDirectory(prefix="ghidra_proj_") as workdir:
            script_dir = os.path.join(workdir, "scripts")
            os.makedirs(script_dir, exist_ok=True)
            script_path = os.path.join(script_dir, "FindXorKey.py")
            with open(script_path, "w", encoding="utf-8") as f:
                f.write(GHIDRA_SCRIPT)

            cmd = [
                analyze,
                workdir, "proj",
                "-import", path,
                "-scriptPath", script_dir,
                "-postScript", "FindXorKey.py",
            ]
            out = subprocess.check_output(cmd, text=True, errors="ignore", timeout=120)
    except subprocess.SubprocessError as e:
        print(f"[!] Ghidra 执行失败: {e}", file=sys.stderr)
        sys.exit(1)
    except OSError as e:
        print(f"[!] 文件操作失败: {e}", file=sys.stderr)
        sys.exit(1)

    # 从输出中提取 XOR key
    m = re.search(r"FOUND_KEY=0x([0-9a-fA-F]+)", out)
    if m and m.group(1).lower() == EXPECTED_KEY:
        print(f"[+] 检测到 XOR key = 0x{EXPECTED_KEY}")
        print(FLAG)
    else:
        print("[!] 未能从 Ghidra 输出中解析到 key，请手动检查：", file=sys.stderr)
        print(out, file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
